packages/lib/hooks/useTypedQuery.ts

```diff
@@ -0,0 +1,59 @@
+import { useRouter } from "next/router";
+import { z } from "zod";
+
+export function useTypedQuery<T extends z.Schema>(schema: T) {
+  type InferedSchema = z.infer<typeof schema>;
+  type SchemaKeys = keyof InferedSchema;
+  const { query: unparsedQuery, ...router } = useRouter();
+  const parsedQuery = schema.parse(unparsedQuery) as InferedSchema;
+
+  // Set the query based on schema values
+  function setQuery<J extends SchemaKeys>(key: J, value: InferedSchema[J]) {
+    router.replace({ query: { ...parsedQuery, [key]: value } }, undefined, { shallow: true });
+  }
+
+  // Delete a key from the query
+  function removeByKey(key: SchemaKeys) {
+    const { [key]: _, ...newQuery } = parsedQuery;
+    router.replace({ query: newQuery }, undefined, { shallow: true });
+  }
+
+  // push item to existing key
+  function pushItemToKey<J extends SchemaKeys>(
+    key: J,
+    value: InferedSchema[J] extends Array<unknown>
+      ? InferedSchema[J][0]
+      : InferedSchema[J] extends Array<unknown> | undefined
+      ? NonNullable<InferedSchema[J]>[0]
+      : InferedSchema[J]
+  ) {
```

Comment:
In short 
if its array use the first type if not use the type itsself 

oh and also checking if its undefined